cart.py

```diff
@@ -1,40 +1,67 @@
+import re
+from typing import List
+from product import Product
 
 
-from product import  Product
 class Cart:
+    def __init__(self, products: list):
+        self.products = products
+        self.basket: List[Product] = []
 
-    def __init__(self):
-        self.products = []
+    def add_product_to_basket(self, new_product_name: str):
+        try:
+            if not re.match("^[A-Za-zĄąĆćĘęŁłŃńÓóŚśŹźŻż ]+$", new_product_name):
+                print("\nNazwa produktu nie może być liczbą lub znakiem specjalnym.")
+                return
 
-    def add_product(self, product: Product):
-        if isinstance(product, Product):
-          self.products.append(product)
+            product = next(
+                (p for p in self.products if p["name_tag"] == new_product_name), None
+            )
+            if product:
+                self.basket.append(product)
+                print(f"\nProdukt {new_product_name} został dodany!")
+            else:
+                print(f"\nProdukt {new_product_name} nie został znaleziony!")
+        except AttributeError as e:
+            print(f"\nError: {e}.")
 
-        else:
-            raise TypeError("This has to be an instance of Product")
+    def remove_product(self, new_product_name: str):
+        try:
+            if not re.match("^[A-Za-zĄąĆćĘęŁłŃńÓóŚśŹźŻż ]+$", new_product_name):
+                print("\nNazwa produktu nie może być liczbą lub znakiem specjalnym.")
+                return
 
-    def remove_product(self, index: int):
-        if 0 <= index:
-            removed_product = self.products.pop(index)  # It'll delete an item with selected index
-            print(f"Produkt {removed_product.name_tag} został usunięty.")
-        else:
-            print("Nieprawidłowy indeks!")
+            product = next(
+                (p for p in self.basket if p["name_tag"] == new_product_name), None
+            )
+            if product:
+                self.basket.remove(product)
+                print(f"\nProdukt {new_product_name} został usunięty!")
+            else:
+                print(f"\nProdukt {new_product_name} nie został znaleziony!")
+        except AttributeError as e:
+            print(f"\nError: {e}.")
 
     def total_price(self):
         total_price_of_products = {sum(product.price for product in self.products)}
-        return f"Twój koszyk jest warty: {total_price_of_products} PLN" if len(self.products) > 0 else ""
+        return (
+            f"Twój koszyk jest warty: {total_price_of_products} PLN"
+            if len(self.products) > 0
+            else ""
+        )
 
     def show_cart(self):
-        if len(self.products) == 0:
+        if len(self.basket) == 0:
             print("Twój koszyk jest pusty")
         else:
-            for product in self.products:
-                print(product)
+            for product in self.basket:
+                print(f"- {product['name_tag']}, cena: {product['price']} PLN\n")
```

Comment:
Proponuje użycie formatowania ":.2f" aby zamiast wyświetlania ceny z jedną liczbą po przecinku wyświetlały się dwie cyfry po przecinku :)
przed: print(f"- {product['name_tag']}, cena: {product['price']} PLN\n")
po: print(f"- {product['name_tag']}, cena: {product['price']:.2f} PLN\n")

Reply:
Tak, jak omawialiśmy to nie jest część zadania. Należy poprowić to w kolejnym.